### PR TITLE
combine multiple case-updates in one form

### PR DESCRIPTION
**Pull request by sravfeyn**
Attempting one piece of https://dimagi-dev.atlassian.net/browse/IIO-502

### CODE REVIEW

File: corehq/apps/callcenter/sync_user_case.py [hunk 1]
```diff
@@ -57,18 +61,18 @@ def update_user_case(self, case, fields, close):
...
         )
-        self._submit_case_block(caseblock, "update_user_case")
+        self._case_block_to_submit = caseblock
         self._user_case_changed(fields)
 
     def _user_case_changed(self, fields):
         field_names = list(fields)
         if _domain_has_new_fields(self.domain, field_names):
-            add_inferred_export_properties.delay(
+            self._tasks_to_trigger = (add_inferred_export_properties, (
```
[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/apps/callcenter/sync_user_case.py]: Doesn't really apply here

File: corehq/apps/callcenter/sync_user_case.py [hunk 2]
```diff
@@ -214,9 +219,31 @@ def _call_center_location_owner(user, ancestor_level):
...
+    first time.
+    """
+    with CriticalSection(get_sync_lock_key(user._id)):
+        helpers = [get_sync_usercase_helper(), get_call_center_case_helper()]
+        case_blocks = [h._case_block_to_submit for h in helpers if h]
+        case_blocks = [cb.as_text() for cb in case_blocks if cb]
+        submit_case_blocks(case_blocks, user.domain, device_id="sync_user_case")
+        for helper in helpers:
+            for task, task_args in helper._tasks_to_trigger:
+                task.delay(task_args))
```
Comment (snopoke) [inline corehq/apps/callcenter/sync_user_case.py]: how about moving this to a function on the helper that can be called here and in `commit` that would keep it encapsulated in the class

File: corehq/apps/callcenter/sync_user_case.py [hunk 3]
```diff
@@ -214,9 +219,31 @@ def _call_center_location_owner(user, ancestor_level):
...
+    """
+    Each time a CommCareUser is saved this method gets called and creates or updates
+    a case associated with the user with the user's details.
+
+    This is also called to create user cases when the usercase is used for the
+    first time.
+    """
+    with CriticalSection(get_sync_lock_key(user._id)):
+        helpers = [get_sync_usercase_helper(), get_call_center_case_helper()]
+        case_blocks = [h._case_block_to_submit for h in helpers if h]
```
Comment (snopoke) [inline corehq/apps/callcenter/sync_user_case.py]:
```suggestion
        case_blocks = itertools.chain([h.get_case_blocks() for h in helpers if h])
```

Then you can move the next line into the `get_case_blocks` function and you won't be accessing a private field

**Review by snopoke — CHANGES_REQUESTED** [review]

**Review by snopoke — APPROVED** [review]

File: corehq/apps/callcenter/sync_user_case.py [hunk 4]
```diff
@@ -171,12 +175,22 @@ def _not_same(val1, val2):
...
+
 def sync_call_center_user_case(user):
+    helper = get_call_center_case_helper(user)
+    if helper:
+        helper.commit()
+
+
+def get_call_center_case_helper(user):
     config = user.project.call_center_config
     if config.enabled and config.config_is_valid():
```
Comment (millerdev) [inline corehq/apps/callcenter/sync_user_case.py]: There's a smell in having this function implicitly return `None` when this condition evaluates to false. In the process of trying to think of a better solution I ended up doing some refactoring (https://github.com/dimagi/commcare-hq/pull/26118). I think the comments on that PR demonstrate why this one is not safe to merge.

**Review by millerdev — CHANGES_REQUESTED** [review]
Don't know why stickler got so confused on this PR. I deleted a bunch of comments because they were making it very hard to review.

**Comment by sravfeyn** [comment]
Hi @millerdev I have updated the code to make it pretty much logically consistent with the way it was, except in the combined case. Apart from the DRYness of commit/commit_multiple, is there any issue that you see related to the locks in this PR anymore?

**Comment by millerdev** [comment]
> Apart from the DRYness of commit/commit_multiple, is there any issue that you see related to the locks in this PR anymore?

Yes. It is using different values for critical section lock keys, which is a significant change from the original design. Without a much deeper analysis I can only conclude that such a change is not safe. `commit_multiple` is especially unsafe as explained in https://github.com/dimagi/commcare-hq/pull/26118#discussion_r354304346.

**Comment by snopoke** [comment]
@millerdev I think your original concern was that things were not happening inside the lock, this has now been changed. The comment you reference says that locking used to be on `owner_id` which is not true, the locking previously was on `case_type, user_id`. I think changing that to `user_id` is safe.

@sravfeyn is there a reason not to include the changes from https://github.com/dimagi/commcare-hq/pull/26118?

**Comment by sravfeyn** [comment]
> @sravfeyn is there a reason not to include the changes from #26118?

I have now included some part of that PR (yielding instead of returning None). The only thing I didn't include is combining `commit` and `commit_multiple`. Personally, I prefer the ease of readability of separate functions over the little DRYness when combined. But happy to change if you both feel that's not true.

**Comment by snopoke** [comment]
Personally I think the combined version is very readable - especially if assertion logic is simplified (I think you only need to check that the there is only 1 user ID being changed (change `_UserCaseHelper.owner_id` to `_UserCaseHelper.user_id` since `owner_id` is not used)

**Comment by millerdev** [comment]
I got confused while writing #26118 because multiple different lock keys were being used (see https://github.com/dimagi/commcare-hq/pull/26118#discussion_r356786413). Right now the lock key in this PR is still different than it was before:

```diff
- old key: 'user_case_%s_for_%s' % (case_type, commcare_user._id)
+ new key: "sync_user_case_for_%s" % user_id
```

I do not think this change is necessary and is possibly unsafe for a sort period during/after deploy. Maybe that's not a big deal?

I think I can sign off on this once the changes suggested in https://github.com/dimagi/commcare-hq/pull/26105#issuecomment-564505371 have been implemented.

**Comment by sravfeyn** [comment]
@snopoke @millerdev Unified

**Review by snopoke — APPROVED** [review]

**Comment by millerdev** [comment]
@snopoke does the lock key change not seem like an issue to you?

**Comment by snopoke** [comment]
@millerdev I'm not concerned about the locking changes given the low frequency of these and that there is already locking at the case level + the fact that repeated calls will almost certainly not conflict with each other in terms of the changes that are being made

File: corehq/apps/callcenter/sync_user_case.py [hunk 5]
```diff
@@ -57,18 +69,17 @@ def update_user_case(self, case, fields, close):
             close=close,
             update=fields
         )
```
Comment (snopoke) [inline corehq/apps/callcenter/sync_user_case.py]:
```suggestion
        ))
```
*** Error compiling 'corehq/apps/callcenter/sync_user_case.py'...
  File "corehq/apps/callcenter/sync_user_case.py", line 72
    self._user_case_changed(fields)
       ^
SyntaxError: invalid syntax

**Review by millerdev — APPROVED** [review]

**Pull request merged**